os/arch/arm/src/s5j/s5j_ticksuppress.c

```diff
@@ -0,0 +1,273 @@
+ *
+ * The timer counts at the rate SYSCLK_FREQUENCY as defined in the board.h
+ * header file.
+ */
+#define SYSTICK_RELOAD ((SYSCLK_FREQUENCY / CLK_TCK) - 1)
+
+#define MAX_TCNT UINT32_MAX	/* Max 32-bit timer-tick value */
+
+/****************************************************************************
+ * Type Definitions
+ ****************************************************************************/
+enum {
+	TICKSUPPRESS_DISABLE,
+	TICKSUPPRESS_ENABLE
+} timertick_mode;
+
+enum {
+	START,
+	READ,
+	WRITE,
+	STOP
+} timer_states;
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+static uint8_t g_tick_mode;	/* timer-tick mode  */
+
+static uint32_t last_event = 0;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static uint32_t s5j_get_next_event(void)
+{
+	return (uint32_t)wd_getdelay();
+}
+
+static uint32_t s5j_timer_setup(int state, uint32_t value)
+{
+	switch (state) {
+		case START:
+			modifyreg32(S5J_RTC_RTCCON, RTC_RTCCON_TICCKSEL0_MASK, RTC_RTCCON_TICKEN0_ENABLE | RTC_RTCCON_TICCKSEL0_32768HZ);
+			break;
+		case READ:
+			return getreg32(S5J_RTC_CURTICCNT0);
+		case WRITE:
+			putreg32(SYSTICK_RELOAD * value, S5J_RTC_TICCNT0);
+			break;
+		case STOP:
+			/* Clear interrupt pending */
+
+			putreg32(RTC_INTP_TIMETIC0, S5J_RTC_INTP);
+
+			/* Disable timertick */
+
+			modifyreg32(S5J_RTC_RTCCON, RTC_RTCCON_TICKEN0_ENABLE, 0);
+			break;
+		default:
+			/* Shouldn't reach here */
+			DEBUGPANIC();
+			break;
+	}
+
+	return 0;
+}
+
+static void s5j_set_next_event(uint32_t next_event)
+{
+	/* Disable timer tick */
+
+	s5j_timer_setup(STOP, 0);
+
+	/* Configure the RTC timetick to generate interrupt on next event */
+
+	s5j_timer_setup(WRITE, next_event);
+
+	/* Enable timer tick */
+
+	s5j_timer_setup(START, 0);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Function:  up_stop_ticksuppress
+ *
+ * Description:
+ *   Disables tick suppression (re-enables periodic ticks) when,
+ *   1. Scheduled event expires
+ *   2. Context switch takes place on an external event (interrupt)
+ *
+ ****************************************************************************/
+
+void up_stop_ticksuppress(void)
+{
+	/* If tick suppression is disabled, do nothing! */
+
+	if (g_tick_mode == TICKSUPPRESS_DISABLE) {
+		return;
+	}
+
+	/* If this function called from interrupt context,
+	 * read current value of the counter and update last event value
+	 */
+
+	if (up_interrupt_context()) {
+		last_event = ((uint32_t)s5j_timer_setup(READ, 0) / SYSTICK_RELOAD);
```

Comment:
does s5j provides a overflow counter to hold the number of times, the counter has been overflowed ? If it provides, we may have to consider handling that case as well

Reply:
I'm not sure about such overflow counter in s5j. I will check